### PR TITLE
[3.4] Adding optional revision bump and mark compacted to snapshot restore

### PR DESCRIPTION
**Pull request by wenjiaswe**
3.4 backport of https://github.com/etcd-io/etcd/pull/16029 for https://github.com/etcd-io/etcd/issues/16028

### CODE REVIEW

**Comment by wenjiaswe** [comment]
cc @dusk125 @tjungblu @ahrtr

File: clientv3/snapshot/v3_snapshot.go [hunk 1]
```diff
@@ -417,6 +434,70 @@ func (s *v3Manager) saveDB() error {
...
+func (s *v3Manager) modifyLatestRevision(bumpAmount uint64) error {
+	dbpath := filepath.Join(s.snapDir, "db")
+	be := backend.NewDefaultBackend(dbpath)
+	defer func() {
+		be.ForceCommit()
+		be.Close()
+	}()
+
+	tx := be.BatchTx()
+	defer tx.Unlock()
```
Comment (ahrtr) [inline clientv3/snapshot/v3_snapshot.go:448]:
```suggestion
	tx.Lock()
	defer tx.Unlock()
```

**Comment by ahrtr** [comment]
Please signoff the second commit, thx

**Comment by wenjiaswe** [comment]
> Please signoff the second commit, thx

Thanks for reminding! Keep forgetting...

**Comment by wenjiaswe** [comment]
hmm, this seems to be flaky, 1 out of 3 failed:

```
--- FAIL: TestWatchDelayForEvent (26.97s)
    --- PASS: TestWatchDelayForEvent/NoTLS (6.35s)
    --- PASS: TestWatchDelayForEvent/TLS (6.83s)
    --- PASS: TestWatchDelayForEvent/SeparateHttpNoTLS (6.66s)
    --- FAIL: TestWatchDelayForEvent/SeparateHttpTLS (7.08s)
```

```
--- PASS: TestWatchDelayForEvent (26.01s)
    --- PASS: TestWatchDelayForEvent/NoTLS (6.37s)
    --- PASS: TestWatchDelayForEvent/TLS (6.84s)
    --- PASS: TestWatchDelayForEvent/SeparateHttpNoTLS (6.27s)
    --- PASS: TestWatchDelayForEvent/SeparateHttpTLS (6.53s)
```

**Review by ahrtr — APPROVED** [review]
LGTM

Thanks @wenjiaswe

**Review by serathius — APPROVED** [review]

**Pull request merged**